[FLINK-11353][tests] Port JobManagerHAJobGraphRecoveryITCase to new code base

tillrohrmann opened the pull request:
## What is the purpose of the change

Removed testClientNonDetachedListeningBehaviour because the test was only relevant for
the JobClientActor. Moved testJobPersistencyWhenJobManagerShutdown to
DispatcherHATest#testPersistedJobGraphWhenDispatcherIsShutDown.

## Verifying this change

- Executed ported tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherHATest.java:
```diff
@@ -150,8 +154,22 @@ public void testGrantingRevokingLeadership() throws Exception {
 
 	@Nonnull
-	private HATestingDispatcher createHADispatcher(TestingHighAvailabilityServices highAvailabilityServices, BlockingQueue<DispatcherId> fencingTokens) throws Exception {
+	private HATestingDispatcher createDispatcher(TestingHighAvailabilityServices highAvailabilityServices, Queue<DispatcherId> fencingTokens) throws Exception {
```
I would suggest to group all helper methods after tests.

tillrohrmann replied:
Good point. Will do it.

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherHATest.java:
```diff
@@ -225,13 +243,78 @@ public static JobGraph createNonEmptyJobGraph() {
 
-		HATestingDispatcher(RpcService rpcService, String endpointId, Configuration configuration, HighAvailabilityServices highAvailabilityServices, ResourceManagerGateway resourceManagerGateway, BlobServer blobServer, HeartbeatServices heartbeatServices, JobManagerMetricGroup jobManagerMetricGroup, @Nullable String metricQueryServicePath, ArchivedExecutionGraphStore archivedExecutionGraphStore, JobMa ...
-			super(rpcService, endpointId, configuration, highAvailabilityServices, resourceManagerGateway, blobServer, heartbeatServices, jobManagerMetricGroup, metricQueryServicePath, archivedExecutionGraphStore, jobManagerRunnerFactory, fatalErrorHandler);
+		final HATestingDispatcher dispatcher = createDispatcher(
```
This test does not seem to need `fencingTokens`. 
At the first glance, does `getNumberJobs` has any HA specifics? Could it also belong to `TestingDispatcher`?
If it is true then this test could just use `TestingDispatcher`, although, constructing code would need to be duplicated.

A bit off topic, do we need the `VisibleForTesting` annotation for this method, as it already resides in test code?

tillrohrmann replied:
I think you're right and the test should be moved to `DispatcherTest` because it is not HA specific. Moreover, we should move `getNumberJobs` into `TestingDispatcher`.

The annotation `VisibleForTesting` is not necessary since `HATestingDispatcher` is a testing class. Will remove it while moving the method over to `TestingDispatcher`.

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherHATest.java:
```diff
@@ -225,13 +243,78 @@ public static JobGraph createNonEmptyJobGraph() {
+			dispatcher.shutDown();
+			dispatcher.getTerminationFuture().get();
+
+			assertThat(submittedJobGraphStore.contains(jobGraph.getJobID()), is(true));
```
Will it be deleted if it is not an HA setup?

tillrohrmann replied:
No it won't. But in the non-HA mode, we usually instantiate a `StandaloneSubmittedJobGraphStore` which is a no-op class.

tillrohrmann commented:
Thanks for the review @azagrebin. I've addressed your comments and moved `testPersistedJobGraphWhenDispatcherIsShutDown` to `DispatcherTest` since it is not HA related.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherHATest.java:
```diff
@@ -225,31 +206,82 @@ public static JobGraph createNonEmptyJobGraph() {
+				fencingToken = newFencingToken;
+			}
+
+			if (fencingTokens != null) {
```
True. I will refactor it.
